mysql.go

```
@@ -71,149 +85,76 @@ func (driver *Driver) Initialize(url string) error {
-	// supports multiple statements per query.
-	sqlStmts := bytes.Split(f.Content, []byte(";"))
-
-	for _, sqlStmt := range sqlStmts {
-		sqlStmt = bytes.TrimSpace(sqlStmt)
-		if len(sqlStmt) > 0 {
-			if _, err := tx.Exec(string(sqlStmt)); err != nil {
-				mysqlErr, isErr := err.(*mysql.MySQLError)
-
-				if isErr {
-					re, err := regexp.Compile(`at line ([0-9]+)$`)
-					if err != nil {
-						pipe <- err
-						if err := tx.Rollback(); err != nil {
-							pipe <- err
-						}
-					}
-
-					var lineNo int
-					lineNoRe := re.FindStringSubmatch(mysqlErr.Message)
-					if len(lineNoRe) == 2 {
-						lineNo, err = strconv.Atoi(lineNoRe[1])
-					}
-					if err == nil {
-
-						// get white-space offset
-						// TODO this is broken, because we use sqlStmt instead of f.Content
-						wsLineOffset := 0
-						b := bufio.NewReader(bytes.NewBuffer(sqlStmt))
-						for {
-							line, _, err := b.ReadLine()
-							if err != nil {
-								break
-							}
-							if bytes.TrimSpace(line) == nil {
-								wsLineOffset += 1
-							} else {
-								break
-							}
-						}
-
-						message := mysqlErr.Error()
-						message = re.ReplaceAllString(message, fmt.Sprintf("at line %v", lineNo+wsLineOffset))
-
-						errorPart := file.LinesBeforeAndAfter(sqlStmt, lineNo, 5, 5, true)
-						pipe <- errors.New(fmt.Sprintf("%s\n\n%s", message, string(errorPart)))
-					} else {
-						pipe <- errors.New(mysqlErr.Error())
-					}
-
-					if err := tx.Rollback(); err != nil {
-						pipe <- err
-					}
-
-					return
-				}
-			}
-		}
+	err = migration.exec(drv.db)
+	if err != nil {
+		return err
 	}
 
-	if err := tx.Commit(); err != nil {
-		pipe <- err
-		return
+	versionUpdSQL := "INSERT INTO " + versionsTableName + " (version) VALUES (?)"
+	if f.Direction == direction.Down {
+		versionUpdSQL = "DELETE FROM " + versionsTableName + " WHERE version = ?"
+	}
+	if _, err = drv.versionTx.Exec(versionUpdSQL, f.Version); err != nil {
+		drv.rollbackVersionTx()
```

for the record, if the rollback fails, the error is lost here.